third_party/move/move-model/bytecode/src/astifier.rs

```diff
@@ -1429,59 +1430,544 @@ struct IfElseTransformer<'a> {
+                    &default_loc,
+                    then_branch.clone(),
+                    0,
+                    /*allow_exit*/ true,
+                )
+                .expect("then_branch must be terminated");
+
+            // Make sure the <then_branch> does not refer to the current loop
+            if then_branch.customizable_branches_to(branch_cond) {
+                return None;
+            }
+            then_branch = then_branch.rewrite_loop_nest(-1);
+            let cond = if_else_list.pop().expect("condition is expected");
+            new_body = self.builder.if_else(
+                cond,
+                then_branch,
+                self.builder.seq(&default_loc, vec![else_branch, new_body]),
+            );
+
+            else_branch = if_else_list.pop()?;
+            // get the terminated prefix of <else_branch> or fallback to the original
+            else_branch = self
+                .builder
+                .extract_terminated_prefix(
+                    &default_loc,
+                    else_branch.clone(),
+                    0,
+                    /*allow_exit*/ true,
+                )
+                .unwrap_or(else_branch);
+
+            // make sure the <else_branch> does not refer to the current loop
+            if else_branch.customizable_branches_to(branch_cond) {
+                return None;
+            }
+            // in case the <else_branch> branches to an outer loop
+            else_branch = else_branch.rewrite_loop_nest(-1);
+        }
+
+        // Add the <begin_stmts> to the top of the expression
+        Some(self.builder.seq(&default_loc, vec![else_branch, new_body]))
+    }
+
+    /// Attempts to remove redundant expressions generated during control flow transformations.
+    fn try_remove_redundant_exp(&self, mut exp: Exp) -> Option<Exp> {
+        let removers: [fn(&Self, Exp) -> Option<Exp>; 2] = [
+            Self::try_remove_redundant_seq,
+            Self::try_remove_redundant_code,
+        ];
+
+        let mut changed = false;
+        for remover in removers {
+            if let Some(new_exp) = remover(self, exp.clone()) {
+                exp = new_exp;
+                changed = true;
+            }
+        }
+
+        if changed {
+            Some(exp)
+        } else {
+            None
+        }
+    }
+
+    /// Attempts to remove dangling `seq` expressions produced by control flow transformations.
+    ///
+    /// It converts the following pattern
+    /// ```move
+    ///   seq {
+    ///     stmt1
+    ///     stmt2
+    ///     seq_1 {
+    ///       stmt3
+    ///       stmt4
+    ///     }
+    ///     seq_2 {
+    ///       stmt5
+    ///       stmt6
+    ///     }
+    ///     ...
+    ///   }
+    /// ==>
+    ///   seq {
+    ///     stmt1
+    ///     stmt2
+    ///     stmt3
+    ///     stmt4
+    ///     stmt5
+    ///     stmt6
+    ///     ...
+    ///   }
+    /// ```
+    fn try_remove_redundant_seq(&self, exp: Exp) -> Option<Exp> {
```

Comment:
Could we do this recursively? Otherwise, the fixpoint loop will do this one layer by one layer.

nit: name this function `flatten_sequence`.